File: apps/bots/src/parser/modules/song.ts
import { UserIntegration } from '@tsuwari/spotify';

import { LastFMDbData, LastFmIntegration } from '../../integrations/lastfm.js';
import { SpotifyIntegration } from '../../integrations/spotify.js';
import { VKDBData, VKIntegration } from '../../integrations/vk.js';
import { prisma } from '../../libs/prisma.js';
import { Module } from '../index.js';

export const song: Module = {
  key: 'currentsong',
  handler: async (_, state) => {
    const enabledIntegrations = await prisma.channelIntegration.findMany({
      where: {
        channelId: state.channelId, 
        enabled: true,
      },
      include: {
        integration: true,
      },
    });

    let result = '';

    for (const integration of enabledIntegrations) {
      switch (integration.integration.service) {
        case 'SPOTIFY': {
          if (!integration.accessToken || !integration.refreshToken) continue;
          const instance = await new UserIntegration(state.channelId, SpotifyIntegration, prisma).init();
          if (!instance) continue;
          const song = await instance.getCurrentSong();
          if (song) result = song;
          
          break;
        }
        case 'VK': {
          const data = integration.data as unknown as VKDBData;
          if (!data.userId) continue;
          const song = await VKIntegration.fetchSong(data.userId);
          if (song) result = song;
          
          break;
        }
        case 'LASTFM': {
          const data = integration.data as unknown as LastFMDbData;
          if (!data.username) continue;
          const song = await LastFmIntegration.fetchSong(data.username);
          if (song) {
            result = song;
          }
          break;
        }

        default:
          break;
      }
    }

    if (result && result.length) {
      return result;
    } else {
      return 'Not playing any song.';
    }
  },
};

File: apps/parser/internal/variables/song/song.go
package song

import (
	"context"
	"fmt"
	"github.com/redis/go-redis/v9"
	"github.com/samber/lo"
	"github.com/satont/tsuwari/apps/parser/internal/types"
	model "github.com/satont/tsuwari/libs/gomodels"
	spotify "github.com/satont/tsuwari/libs/integrations/spotify"
)

var Song = &types.Variable{
	Name:        "currentsong",
	Description: lo.ToPtr("Print current played song from Spotify, Last.fm, e.t.c, and also from song requests."),
	Handler: func(ctx context.Context, parseCtx *types.VariableParseContext, variableData *types.VariableData) (*types.VariableHandlerResult, error) {
		result := &types.VariableHandlerResult{}

		integrations := parseCtx.Cacher.GetEnabledChannelIntegrations(ctx)

		integrations = lo.Filter(
			integrations,
			func(integration *model.ChannelsIntegrations, _ int) bool {
				switch integration.Integration.Service {
				case "SPOTIFY", "VK", "LASTFM":
					return integration.Enabled
				default:
					return false
				}
			},
		)

		lastFmIntegration, ok := lo.Find(
			integrations,
			func(integration *model.ChannelsIntegrations) bool {
				return integration.Integration.Service == "LASTFM"
			},
		)

		var lfm *lastFm
		if ok {
			lfm = newLastfm(lastFmIntegration)
		}

		spotifyIntegration, ok := lo.Find(
			integrations,
			func(integration *model.ChannelsIntegrations) bool {
				return integration.Integration.Service == "SPOTIFY"
			},
		)
		var spoti *spotify.Spotify
		if ok {
			spoti = spotify.New(spotifyIntegration, parseCtx.Services.Gorm)
		}

		vkIntegration, ok := lo.Find(
			integrations,
			func(integration *model.ChannelsIntegrations) bool {
				return integration.Integration.Service == "VK"
			},
		)
		var vk *vkService
		if ok {
			vk = newVk(vkIntegration)
		}

		integrationsForFetch := lo.Map(
			integrations,
			func(integration *model.ChannelsIntegrations, _ int) model.IntegrationService {
				return integration.Integration.Service
			},
		)

		integrationsForFetch = append(integrationsForFetch, model.IntegrationService("YOUTUBE_SR"))

	checkServices:
		for _, integration := range integrationsForFetch {
			switch integration {
			case model.IntegrationServiceSpotify:
				if spoti == nil {
					continue
				}
				track := spoti.GetTrack()
				if track != nil {
					result.Result = *track
					break checkServices
				}
			case model.IntegrationServiceLastfm:
				if lfm == nil {
					continue
				}

				track := lfm.GetTrack()

				if track != nil {
					result.Result = *track
					break checkServices
				}
			case model.IntegrationServiceVK:
				if vk == nil {
					continue
				}
				track := vk.GetTrack(ctx)
				if track != nil {
					result.Result = *track
					break checkServices
				}
			case "YOUTUBE_SR":
				redisData, err := parseCtx.Services.Redis.Get(
					context.Background(),
					fmt.Sprintf("songrequests:youtube:%s:currentPlaying", parseCtx.Channel.ID),
				).Result()
				if err == redis.Nil {
					continue
				}
				if err != nil {
					parseCtx.Services.Logger.Sugar().Error(err)
					continue
				}
				song := model.RequestedSong{}
				if err = parseCtx.Services.Gorm.
					WithContext(ctx).
					Where("id = ?", redisData).
					First(&song).Error; err != nil {
					fmt.Println("song nog found", err)
					continue
				}

				result.Result = fmt.Sprintf(
					`"%s" youtu.be/%s requested by @%s`,
					song.Title,
					song.VideoID,
					song.OrderedByName,
				)
				break checkServices
			}
		}

		return result, nil
	},
}
